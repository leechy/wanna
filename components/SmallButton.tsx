// hooks
import { useThemeColor } from '@/hooks/useThemeColor';

// components
import { StyleSheet, Text, View, type ViewStyle } from 'react-native';
import { PressableArea } from './PressableArea';

// types
import { ReactNode } from 'react';
import { SvgProps } from 'react-native-svg';

export type SmallButtonProps = {
  style?: ViewStyle;
  title?: String;
  icon?: React.FC<SvgProps>;
  onPress?: () => void;
  color?: string;
  children?: ReactNode;
};

export default function SmallButton({ title, icon, style, onPress, color, children }: SmallButtonProps) {
  const smallButtonText = useThemeColor({}, 'touchable');
  const smallButtonBackground = useThemeColor({}, 'smallButtonBackground');
  const inactiveColor = useThemeColor({}, 'inactive');

  const Icon = icon || (() => <></>);

  function getContents() {
    return (
      <>
        <Icon width={24} height={24} color={color || smallButtonText} style={styles.icon} />
        {title ? (
          <Text
            style={[
              {
                color: color || smallButtonText,
              },
              styles.text,
            ]}
          >
            {title}
          </Text>
        ) : (
          children
        )}
      </>
    );
  }

  return (
    <View
      style={[
        styles.container,
        style,
        { backgroundColor: smallButtonBackground, borderColor: color ? color + '80' : inactiveColor },
      ]}
    >
      {onPress ? (
        <PressableArea style={styles.button} onPress={onPress}>
          {getContents()}
        </PressableArea>
      ) : (
        <View style={styles.button}>{getContents()}</View>
      )}
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    borderRadius: 8,
    overflow: 'hidden',
    borderWidth: 1,
  },
  background: {
    borderRadius: 8,
  },
  button: {
    flexDirection: 'row',
    alignItems: 'center',
    minHeight: 24,

    borderRadius: 8,
    paddingTop: 1,
    paddingHorizontal: 12,
  },
  icon: {
    marginRight: 4,
    marginLeft: -8,
    marginBottom: 2,
  },
  text: {
    fontFamily: 'Montserrat',
    fontSize: 14,
    fontWeight: '400',
    lineHeight: 16,
    textAlign: 'center',
  },
});
