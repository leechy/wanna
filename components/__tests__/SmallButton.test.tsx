import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import SmallButton from '../SmallButton';

describe('SmallButton', () => {
  it('renders correctly', () => {
    const { getByText } = render(
      <SmallButton title="Click me" onPress={() => {}} />
    );
    expect(getByText('Click me')).toBeTruthy();
  });

  it('calls onPress when pressed', () => {
    const onPressMock = jest.fn();
    const { getByText } = render(
      <SmallButton title="Click me" onPress={onPressMock} />
    );
    const button = getByText('Click me');

    fireEvent.press(button);
    expect(onPressMock).toHaveBeenCalled();
  });

  it('matches snapshot', () => {
    const { toJSON } = render(
      <SmallButton title="Click me" onPress={() => {}} />
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
