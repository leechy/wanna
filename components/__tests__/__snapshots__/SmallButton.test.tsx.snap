// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SmallButton matches snapshot 1`] = `
<View
  style={
    [
      {
        "borderRadius": 8,
        "borderWidth": 1,
        "overflow": "hidden",
      },
      undefined,
      {
        "backgroundColor": "#ffffff",
        "borderColor": "#999999",
      },
    ]
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "borderRadius": 8,
          "flexDirection": "row",
          "paddingHorizontal": 12,
          "paddingTop": 1,
        },
        {
          "backgroundColor": "transparent",
          "opacity": 1,
        },
      ]
    }
  >
    <Text
      style={
        [
          {
            "color": "#666666",
          },
          {
            "fontFamily": "Montserrat",
            "fontSize": 14,
            "fontWeight": "400",
            "lineHeight": 16,
            "textAlign": "center",
          },
        ]
      }
    >
      Click me
    </Text>
  </View>
</View>
`;
